### ISSUE TITLE
Improvement: Merge individual rule's configuration when merging config yml files

### ISSUE THREAD

**Issue by gretzki** [realm/SwiftLint#2058]
When merging Swiftlint configuration files, the set of rules are merged according to the rulesMode flag, but not the configuration of the rules that are carried over. (Since #1687)

**Just for clarification, let me define the following terms:**
Having two configurations for one rule, one at the current directory - subsequently called `nested configuration` - and one at a higher level - subsequently called `parent configuration`. When merging the rules configuration for the current directory, the result is subsequently called the `result configuration`.

**Assume the following example:**
The parent configuration specifies:

> cyclomatic_complexity:
>   warning: 13
>   error: 14
>   ignores_case_statements: true

and the nested configuration specifies:

> cyclomatic_complexity:
>   warning: 5

 **My assumption:**
The result configuration should have:

> cyclomatic_complexity:
>   warning: 5
>   error: 14
>   ignores_case_statements: true

**Current implementation**
The result configuration depends if it is an optIn rule. If it was, the result would specify the  warning threshold of the nested configuration, but default values for the other settings:

> cyclomatic_complexity:
>   warning: 5
>   error: 20
>   ignores_case_statements: false

Since it is not an optIn rule, the parent configuration is taken over and the result configuration has

> cyclomatic_complexity:
>   warning: 13
>   error: 14
>   ignores_case_statements: true 

**My proposition**
Both cases are counterintuitive (at least for me). Therefore, I propose the following:
The result configuration should always contain the rule's configuration of the parent, overriden by what has been set by the nested configuration explicitly.

If this is the expected behaviour one can agree on,  I'm more than happy to publish a pull request for the proposed behaviour. Btw. kudos for the excellent work on #1687.

**Comment by gretzki** [comment]
I've opened a feature branch as a proof of concept and to make the changes more easily traceable: #2059 

In summary, the changes are:
1. Rules are initially configured via init(configuration:) in Rule.swift when parsing yml files. There, the set of explicitly set values (compared to default values) can be determined and stored.
WiP branch still uses another place to store the explicitly set values.

2. Configuration+Merging.swift is extended to merge rule's configurations too, when merging the set of rules considering disabled sets and optIn sets. To merge a rule's configuration, a method called `overrideConfiguration(configuration:)` is invoked.
WiP branch only extends the ruleMode .default.

3. Rule is extended with the function `overrideConfiguration(withRule secondRule: Rule) -> Rule`.
It works as follows:
- lookup own RuleConfiguration property A
- lookup second rule's RuleConfiguration property B
- call A.apply(B)

Checking that A and B are of same type.
Also using an enclosing try..catch block to stay backwards compatible for rules not implementing the new behaviour yet.

4. Extending concrete implementations of RuleConfiguration, so that the apply(...) method can be invoked with parameters of same type.
WIP branch implements this exemplary in CyclomaticComplexityConfiguration.swift.


The feature branch is working as expected, i.e. it merges the rule's configurations as proposed above. There are still some issues that need to be addressed (but does not affected the functioning).

ToDos:
- Create an unit test to cover the new behaviour
- Change implementation to not depend on extending all RuleConfiguration implementations 
- Fix breaking test „testNestedWhitelistedRules“
- Remove ugly hack to check if a Rule conforms to ConfigurationProviderRule (in Rule+Merging.swift)

**Comment by GeroHerkenrath** [comment]
Hey @gretzki! What a coincidence, I just came across the same problem. Nice to see you're still working with ios (or mac?) things. Kudos for the proposed fix, I will gladly use it with the next swiftlint release! 👍

**Comment by grantneufeld** [comment]
This problem seems to still be occurring.

My `project/.swiftlint.yml` file just has some `disabled_rules` and `opt_in_rules`; no rule-specific configurations, and no mention of `identifier_name`.

My `project/App/models/.swiftlint.yml` file tries to override the `identifier_name`, changing `min_length` from the default of 3 to 2:

```yaml
identifier_name:
  min_length:
    warning:
      2
```

That fails to allow 2 character length identifiers in `project/App/models/Item.swift`, producing an error when it finds `var id: UUID` in my code: “Identifier Name Violation: Variable name should be between 3 and 40 characters long: 'id' (identifier_name)” (Now, maybe I really shouldn’t be using `id` like this here, but bear with me…)

If I copy the code from `project/App/models/.swiftlint.yml` into `project/.swiftlint.yml`, it does allow the 2 character length identifiers (but for the whole project). So, I know that’s the right format for the configuration I’m after. It’s just failing to load it from the config file in the sub-directory.

I also tried just outright disabling the rule in `project/App/models/.swiftlint.yml`:

```yaml
disabled_rules:
  - identifier_name
```

and that worked as expected (but is not quite what I was after, as it would allow things like `var x = 0`). So that confirms that the config in the sub-directory is being loaded, just not the configuration of individual rules in it.
